src/Raven.Server/Documents/AI/Embeddings/QueryEmbeddingsCacher.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Server.Background;
+using Raven.Server.Documents.TransactionMerger.Commands;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Server;
+
+namespace Raven.Server.Documents.AI.Embeddings;
+
+public class QueryEmbeddingsCacher : BackgroundWorkBase
+{
+    private readonly DocumentDatabase _database;
+
+    private readonly ConcurrentQueue<EmbeddingGenerationItem> _embeddingsQueue;
+    private readonly AsyncManualResetEvent _mre;
+    private int _approximateCount;
+    public QueryEmbeddingsCacher(DocumentDatabase database, CancellationToken shutdown) : base(database.Name, database.Loggers.GetLogger<QueryEmbeddingsCacher>(), shutdown)
+    {
+        _database = database;
+        _embeddingsQueue = new ConcurrentQueue<EmbeddingGenerationItem>();
+        _mre = new AsyncManualResetEvent();
+    }
+
+    protected override async Task DoWork()
+    {
+        while (CancellationToken.IsCancellationRequested == false)
+        {
+            await _mre.WaitAsync(CancellationToken);
+
+            _mre.Reset();
+
+            var mightBeMore = CacheEnqueuedEmbeddings();
+            
+            if (mightBeMore)
+                _mre.Set();
+        }
+    }
+
+    internal bool CacheEnqueuedEmbeddings()
+    {
+        var mightBeMore = false;
+        
+        var maxBatchSize = _database.Configuration.Ai.QueryEmbeddingsGenerationMaxCacheBatchSize;
+
+        var payload = new List<EmbeddingGenerationItem>(Math.Min(_approximateCount, maxBatchSize));
+
+        while (_embeddingsQueue.TryDequeue(out var item))
+        {
+            CancellationToken.ThrowIfCancellationRequested();
+                
+            Interlocked.Decrement(ref _approximateCount);
+
+            payload.Add(item);
+
+            if (payload.Count >= maxBatchSize)
+            {
+                mightBeMore = true;
+                break;
+            }
+        }
+
+        if (payload.Count == 0)
+            return mightBeMore;
+
+        var putEmbeddingsCommand = new PutQueryEmbeddingsCommand(payload, _database);
+
+        _database.TxMerger.EnqueueSync(putEmbeddingsCommand);
+
+        return mightBeMore;
+    }
+
+    public void EnqueueEmbeddingsToCache(List<EmbeddingGenerationItem> embeddings)
+    {
+        foreach (EmbeddingGenerationItem item in embeddings)
+        {
+            _embeddingsQueue.Enqueue(item);
+
+            Interlocked.Increment(ref _approximateCount);
```

Comment:
This should be done outside of the loop, only once.

Reply:
Outside the loop it would need to be done like:

```
while (true)
{
    var old = Volatile.Read(ref _approximateCount);
    if (Interlocked.CompareExchange(ref _approximateCount, old + embeddings.Count, old) == old)
        break;
}
```

I'm not sure if it's worth it (way more complex code). It's not going to be performance critical code.

Reply:
`Interlocked.Add(ref _approximateCount, embeddings.Count);` does that in a single instruction.

```
lock add [rbx], rcx
```

Versus this for `Interlocked.Increment()`

```
lock inc qword ptr [rcx]
```

Reply:
done